app/models/state_changers.rb

```diff
@@ -58,6 +58,23 @@ def self.lookup_for(purpose_uuid)
     details[:state_changer_class].constantize
   end
 
+  # The tube rack state changer is used by TubeRacks.
+  # It contains tubes.
+  class TubeRackStateChanger < DefaultStateChanger
+    # TubeRacks contain tubes
+    def contents_for(_target_state)
+      return nil unless FILTER_FAILS_ON.include?(target_state)
+
+      # determine list of tubes requiring the state change
+      tubes_locations_filtered = labware.tubes.reject { |t| t.state == 'failed' }.map(&:racked_tube.coordinate)
```

Comment:
Not sure about the `&:racked_tube`.

A tube can have many `racked_tubes`, can't it, because it can be in many tube racks over time. You could do:

```
labware.racked_tubes.reject { |rt| rt.tube.state == 'failed' }.map(&:coordinate)
```
Would that work?

Reply:
Agreed